vm/core/hash.go

```diff
@@ -14,14 +15,25 @@ func SigningBody(genesis, tx []byte) []byte {
 	return full
 }
 
-// ComputePrincipal address as the last 20 bytes from blake3(scale(template || args)).
-func ComputePrincipal(template Address, args scale.Encodable) Address {
+// ComputePrincipal address as the last 24 bytes of Hash(template || blob).
+// See https://github.com/spacemeshos/go-spacemesh/issues/6420 for more details.
+func ComputePrincipalFromBlob(template types.Address, blob []byte) Address {
 	hasher := hash.GetHasher()
 	defer hash.PutHasher(hasher)
-	encoder := scale.NewEncoder(hasher)
-	template.EncodeScale(encoder)
-	args.EncodeScale(encoder)
+	hasher.Write(template[:])
+	hasher.Write(blob)
 	sum := hasher.Sum(nil)
-	rst := types.GenerateAddress(sum[12:])
-	return rst
+	return types.GenerateAddress(sum)
+}
+
+func ComputePrincipalFromPubkey(template types.Address, pubkey signing.PublicKey) (Address, error) {
+	// construct and encode the blob, which is a SCALE-encoded Athena wallet template instance
+	blob, err := scale.Marshal(struct {
+		Nonce, Balance uint64
```

Comment:
Why do we need `Balance` here?
IMO it looks weird to add balance as a spawn argument for the Wallet account type, it's not a vault or some token smart-contract (total issuance)...

Reply:
Btw, also I guess such `Nonce` field should be a part of payload of spawn tx, and it's the different nonce from the "tx counter"

Reply:
yes, as discussed in Slack, nonce and balance should be removed -- not just from here, but from the VM side as well.